### PR TITLE
refactor: wire in model service bootstrap worker

### PR DESCRIPTION
**Pull request by tlm**
This commit wires in the model service(s) to the bootstrap worker replacing calls to Mongo state for model information. This allows us to keep transitioning off of Mongo and over to dqlite.
    
This change is functionally equivalent with the source of model information being changed.

This work is being done as part of bootstrapping authorised keys. We required the model service(s) to perform this function correctly.

As part of this PR we have also removed manual environ creation using domain services and opted to use the provider tracker available in the manifold.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

1. Run unit tests for `internal/worker/bootstrap`
2. Bootstrap a local controller and switch to the controller model. Confirm the controller charm is health and deployed with reported agent versions.

## Documentation changes

N/A

## Links

**Jira card:** JUJU-6478

### CODE REVIEW

File: internal/worker/bootstrap/addressfinder.go [hunk 1]
```diff
@@ -49,7 +51,18 @@ func CAASBootstrapAddresses(ctx context.Context, env environs.Environ, bootstrap
 
 // IAASBootstrapAddressFinder returns the bootstrap addresses for IAAS.
 func IAASBootstrapAddressFinder(ctx context.Context, config BootstrapAddressesConfig) (network.ProviderAddresses, error) {
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/addressfinder.go:53]: This could just be changed to use the provider tracker, then you wouldn't need this set of changes.

Comment (tlm) [reply internal/worker/bootstrap/addressfinder.go]: Done & Done

File: internal/worker/bootstrap/addressfinder.go [hunk 2]
```diff
@@ -5,110 +5,68 @@ package bootstrap
...
-	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, fmt.Errorf(
+			"cannot get instance %q from instance lister: %w",
+			instanceID, err,
+		)
 	}
-
-	controllerModelConfig, err := controllerModel.Config()
+	addrs, err := instances[0].Addresses(callCtx)
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/addressfinder.go:38]: Add a guard around this raw access so we don't panic by accident.

File: internal/worker/bootstrap/addressfinder_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// TestBootstrapAddressFinder is asserting the happy path of finding an instance
+// addresses via a provider.
+func (s *addressFinderSuite) TestBoostrapAddressFinder(c *gc.C) {
+	ctrl := s.setupMocks(c)
+	defer ctrl.Finish()
+
+	addresses := network.NewMachineAddresses([]string{
+		"2001:0DB8::BEEF:FACE",
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/addressfinder_test.go:66]: I see what you've done here. :wink:

File: internal/worker/bootstrap/manifold.go [hunk 4]
```diff
@@ -286,6 +173,24 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
...
+
+			controllerModel, err := controllerServiceFactory.Model().ControllerModel(ctx)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"cannot get controller model when making bootstrap worker: %w",
+					err,
+				)
+			}
+
+			instanceListerProvider := providertracker.ProviderRunner[environs.InstanceLister](
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/manifold.go:189]: This does all of the heavy lifting. Excellent!

File: internal/worker/bootstrap/worker.go [hunk 5]
```diff
@@ -403,7 +389,14 @@ func (w *bootstrapWorker) scopedContext() (context.Context, context.CancelFunc)
 }
 
 func (w *bootstrapWorker) seedAgentBinary(ctx context.Context, dataDir string) (func(), error) {
-	objectStore, err := w.cfg.ObjectStoreGetter.GetObjectStore(ctx, w.cfg.SystemState.ControllerModelUUID())
+	controllerModel, err := w.cfg.ModelService.ControllerModel(ctx)
```
Comment (manadart) [inline internal/worker/bootstrap/worker.go:392]: Could get this at instantiation. The data never changes.

**Review by manadart — COMMENTED** [review]
Tidies it up nicely.

File: internal/worker/bootstrap/addressfinder.go (hunk 2 above)
Comment (tlm) [reply internal/worker/bootstrap/addressfinder.go:38]: Done & Done

**Review by SimonRichardson — APPROVED** [review]

File: internal/worker/bootstrap/worker.go (hunk 5 above)
Comment (tlm) [reply internal/worker/bootstrap/worker.go]: Done & Done

**Pull request merged**